### PR TITLE
cw310_rom_variant tag

### PR DESCRIPTION
**Pull request by drewmacrae**
Add a check for a cw310_other tag to be added to tests that require bitstreams other than the ones we cache for the cw310. This lets us group the cw310_rom and cw310_test_rom tests to avoid loading bitstreams when we run those tests.

I also increased the timeouts to make it less likely to encounter problems.

### CODE REVIEW

**Review by alphan — COMMENTED** [review]
@drewmacrae I don't think I understand the goal of this change. The tests that are marked as `cw310_other` should actually be `cw310_rom`. I think what this new tag wants to be called is "spliced_locally" or something similar. I agree with your message on slack, this seems to be redundant.

File: ci/scripts/check-bazel-tags.sh
Comment (dmcardle) [inline ci/scripts/check-bazel-tags.sh:27]: s/bistream/bitstream

File: ci/scripts/check-bazel-tags.sh
Comment (dmcardle) [inline ci/scripts/check-bazel-tags.sh:26]: Nit: Could you try to define a function that defines the tag-matching patterns? I think it would improve readability.

This is getting complicated enough that I would support rewriting it in anything other than bash... but maybe another week :)

File: ci/scripts/check-bazel-tags.sh
Comment (dmcardle) [inline ci/scripts/check-bazel-tags.sh:18]: This comment might be a little misleading, could you also change it to say "all targets that depend on updatemem"? Targets like //hw/bitstream/vivado:fpga_cw310 also depend on Vivado.

File: rules/opentitan_test.bzl
Comment (dmcardle) [inline rules/opentitan_test.bzl:326]: I think we'd rather `fail()` than grab the //hw/bitstream:rom for "cw310_other", right?

**Comment by dmcardle** [comment]
My understanding is that these tags are meant to identify categories of bitstreams.

* `cw310_rom`: depends on //hw/bitstream:rom
* `cw310_test_rom`: depends on //hw/bitstream:test_rom
* `cw310_other`: depends on a bitstream, but neither //hw/bitstream:rom nor //hw/bitstream:test_rom

@drewmacrae is the above accurate?

I do think the name could be improved. WDYT about `cw310_rom_variant`? (I think `cw310_rom_spliced` doesn't cut it because technically //hw/bitstream:rom is spliced.)

I left a comment about the default bitstream selected for the new tag, which I think is why @alphan said that the tests marked `cw310_other` should actually be tagged `cw310_rom`.

File: rules/opentitan_test.bzl
Comment (drewmacrae) [reply rules/opentitan_test.bzl:326]: Ack, I think I can fix this. (The default bitstream is looked up and then overwritten so I'll need to construct a dummy here that fails if we use it.)

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh:18]: The rom depends on updatemem but probably shouldn't have the tag... We need to discuss the vivado tag further.

**Comment by drewmacrae** [comment]
This gets 35 minutes of headroom for spliced tests which are currently flaky because of timeouts.

**Comment by drewmacrae** [comment]
> My understanding is that these tags are meant to identify categories of bitstreams.
> 
> * `cw310_rom`: depends on //hw/bitstream:rom
> * `cw310_test_rom`: depends on //hw/bitstream:test_rom
> * `cw310_other`: depends on a bitstream, but neither //hw/bitstream:rom nor //hw/bitstream:test_rom
> 
> @drewmacrae is the above accurate?

That's mostly right.  The dependency graph is more complicated than that.
rom tests depend on 
```
bazel query "kind('bitstream_splice', deps(//third_party/riscv-compliance:I-LBU-01_fpga_cw310_rom))"
//hw/bitstream:gcp_spliced_rom
//hw/bitstream/vivado:fpga_cw310_rom
```

cw310_other tests also may depend on the rom, and you could make one that depends on the test_rom if you wanted to.

I'm unhappy with those queries except insofar as they work.

**Comment by drewmacrae** [comment]
> cw310_rom_variant

I can live with that, thanks @dmcardle

File: rules/opentitan_test.bzl
Comment (drewmacrae) [reply rules/opentitan_test.bzl]: I was wrong, I only needed that before I fixed something else in this file. Thanks!

**Comment by drewmacrae** [comment]
By the way, with this PR, the test_rom_variant tests take 34 min, but they show we're not always loading a bitstream:
```
INFO: Elapsed time: 2049.960s, Critical Path: 88.78s                                                                                                                                                                                                                                                                                                                                                         [196/1981]
INFO: 744 processes: 165 linux-sandbox, 639 local.                                                                                                                                                         
INFO: Build completed successfully, 744 total actions                                                                                                                                                      
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_bad_identifier_a_fpga_cw310_rom_variant PASSED in 17.9s              
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_bad_identifier_b_fpga_cw310_rom_variant PASSED in 1.7s               
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_code_end_unaligned_a_fpga_cw310_rom_variant PASSED in 1.7s          
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_code_end_unaligned_b_fpga_cw310_rom_variant PASSED in 1.8s         
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_code_in_manifest_a_fpga_cw310_rom_variant PASSED in 1.7s             
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_code_in_manifest_b_fpga_cw310_rom_variant PASSED in 1.7s             
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_code_outside_image_a_fpga_cw310_rom_variant PASSED in 1.7s             
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_code_outside_image_b_fpga_cw310_rom_variant PASSED in 1.7s             
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_code_start_unaligned_a_fpga_cw310_rom_variant PASSED in 1.7s 
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_code_start_unaligned_b_fpga_cw310_rom_variant PASSED in 1.8s 
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_empty_code_a_fpga_cw310_rom_variant PASSED in 1.7s                  
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_empty_code_b_fpga_cw310_rom_variant PASSED in 1.7s                  
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_entry_at_code_end_a_fpga_cw310_rom_variant PASSED in 1.7s                 
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_entry_at_code_end_b_fpga_cw310_rom_variant PASSED in 1.7s                 
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_entry_before_code_start_a_fpga_cw310_rom_variant PASSED in 1.7s   
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_entry_before_code_start_b_fpga_cw310_rom_variant PASSED in 1.7s   
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_entry_unaligned_a_fpga_cw310_rom_variant PASSED in 1.7s     
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_entry_unaligned_b_fpga_cw310_rom_variant PASSED in 1.7s    
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_rollback_a_fpga_cw310_rom_variant PASSED in 17.9s            
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_rollback_b_fpga_cw310_rom_variant PASSED in 1.7s            
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_too_large_a_fpga_cw310_rom_variant PASSED in 17.9s          
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_too_large_b_fpga_cw310_rom_variant PASSED in 1.7s           
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_too_small_a_fpga_cw310_rom_variant PASSED in 1.7s 
//sw/device/silicon_creator/rom/e2e:boot_policy_bad_manifest_dev_too_small_b_fpga_cw310_rom_variant PASSED in 1.7s      
```
These tests should always be loading a bitstream right?

File: ci/scripts/check-bazel-tags.sh
Comment (milesdai) [inline ci/scripts/check-bazel-tags.sh:108]: Same as the comment above re: a more explicit error message:

```suggestion
"The target(s) above are tagged with cw310_rom but depend(s) on a bitstream_splice rule other than those used to generate the rom.
Please tag with cw310_*_variant instead if you intended to use another bitstream."
```

File: ci/scripts/check-bazel-tags.sh
Comment (milesdai) [inline ci/scripts/check-bazel-tags.sh:59]: "and" --> "that"?

File: ci/scripts/check-bazel-tags.sh
Comment (milesdai) [inline ci/scripts/check-bazel-tags.sh:38]: Is there any guidance here on whether people should use `verilator` vs `manual`?

**Review by milesdai — APPROVED** [review]
I think this LGTM in general. My comments are all related to user-facing things.

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh:38]: This check isn't strict enough to care which is used. Neither is the one for vivado. We don't have consensus over what should be done, so we shouldn't be stricter than this. I leave it to the intent of the BUILD file's author, but I can add a note regarding why they'd choose manual.

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh:9]: Yes I can put a link there.

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh:76]: ack

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh:85]: I can clean that up a little. The tag is also usually implicitly set based on the target and I should clarify that.

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh:108]: I can clean that up a little. The tag is also usually implicitly set based on the target and I should clarify that.

File: ci/scripts/check-bazel-tags.sh
Comment (dmcardle) [inline ci/scripts/check-bazel-tags.sh:68]: I'm a little confused about this message as well.  How about:

> The targets above depend on a bitstream_splice target, but are not appropriately tagged. Please indicate which kind of bitstream each target depends on by tagging it with one of the following: 'cw310_rom', 'cw310_test_rom', or 'vivado'. Optionally, add the 'manual' tag to exclude the targets from Bazel wildcards.

File: ci/scripts/check-bazel-tags.sh
Comment (dmcardle) [inline ci/scripts/check-bazel-tags.sh:104]: I'm having trouble interpreting this comment relative to the query below.

It looks to me that we're searching for mistagged tests. The check fails if we find a test that depends on a `bitstream_splice` and is tagged `cw310_rom`, provided that the `bitstream_splice` is not used by `//hw/bitstream:rom`.

File: ci/scripts/check-bazel-tags.sh
Comment (dmcardle) [inline ci/scripts/check-bazel-tags.sh:123]: Nit: s/rule/target

**Review by dmcardle — APPROVED** [review]
Mostly LGTM, but I have nits and suggestions for the user-facing strings.

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh:18]: yes

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh:46]: Oh yeah!

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh:75]: That should be "Anything", but good catch. I'll change the query (rdeps includes dependencies so we can't filter below and have to do it above.)

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh:123]: Thanks, target is correct but also overloaded in our context.
There's a var named target that's used in opentitan_functest, and it's used in the system testing context and hardware description context along with device.
I'm going to use "bazel target" here to try to keep things clearer.

Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh:123]: and functest target for the other case in use here

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh:104]: This comment describes why we do the check. The query itself is documented below. I can describe it without using the tags to name the bitstreams.

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh:68]: The cw310_rom and cw310_test_rom tags get added by the functest rule, that's not likely to be an appropriate resolution.

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh:42]: I guess that comma could be a semicolon or a period...?

Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh:42]: BTW (I guess because the edit is on the previous line) this comment shows outdated code in the conversations tab but doesn't show as outdated.

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh]: I'll keep the "Target above" references as is because the context makes it very clear.

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh:42]: @dmcardle I don't know what you're asking for here so I'm blocked on resolving it.

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh]: @dmcardle I don't know what you're asking for here so I'm blocked on resolving it.

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh]: @dmcardle I don't know what you're asking for here so I'm blocked on resolving it.

File: ci/scripts/check-bazel-tags.sh
Comment (dmcardle) [reply ci/scripts/check-bazel-tags.sh:42]: It seemed to me that this guidance could be a little more pointed.

Is the goal to ensure that users without Vivado can run `bazel build //...`? Then tagging these targets with 'manual' is mandatory.

Is the goal to ensure that CI can entirely exclude Vivado-related targets with the 'vivado' tag? Then, tagging with 'vivado' is mandatory.

If both are goals, we should instruct the user to add both tags, right?

Comment (dmcardle) [reply ci/scripts/check-bazel-tags.sh:42]: Sorry, I'm confused, I just realized this comment was about Verilator.

File: ci/scripts/check-bazel-tags.sh
Comment (dmcardle) [reply ci/scripts/check-bazel-tags.sh]: Maybe I'm just tired, but I still find the comment hard to parse. In particular, "with the cw310_rom tag" feels like it's too far away from the verb "group". Also, I wasn't sure what "the cached ROM bitstream" referred to at first. Anyway, same as above, I don't want to block the PR on this.

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh:42]: No worries, the verilator and vivado tags are effectively similar:
Verilator has a longer set of precedents, a failure case that can be mitigated by building a dummy and consensus that intermediate deps shouldn't really be tagged.

There's also not a plan to mitigate the need for the verilator tag like there is for vivado.

The verilator case should be hard to hit: the functest rule adds the tag, so we shouldn't see new targets hit this at all, so the contents of this error message shouldn't be that relevant.

I'll feel free to resolve this and post it as an issue if this PR needs merging, but until then this PRs as good a place as any to sort it out.

The goal is to ensure users without vivado can run `bazel build //... --build_tag_filters=-vivado` With that invocation either manual or vivado will prevent matching and building. I don't like having manual targets with no rdeps because they're likely to reflect dead code. (test_suites are an exception here, kindof)
if the target has rdeps then we have been tagging it manual without objection in the verilator context, and could probably do the same for vivado.)

Other than the rdeps thing, there isn't a consensus and I don't have a notion as to which should be used.

File: ci/scripts/check-bazel-tags.sh
Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh]: > Maybe I'm just tired, but I still find the comment hard to parse. In particular, "with the cw310_rom tag" feels like it's too far away from the verb "group".

Notionally or as written in the sentence?

Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh]: > "the cached ROM bitstream"

What should we call it? I've been calling it cw310_rom which has also caused problems and *.spliced might not even be in use anymore...
Names matter and using the wrong one here will cost more in the long run than fixing it (plus I still don't have approval yet so it's not like I have an excuse to merge it in and force the issue)

"the cw310 bitstream with the ROM spliced into it and hosted in GCP" feels wordy, but if that's what it takes to be clear, I might as well get it right in a comment. We could hyphenate it so people couldn't call it wordy, and because hyphenating long strings is a good schtick.

Comment (drewmacrae) [reply ci/scripts/check-bazel-tags.sh]:
```
# This check ensures cw310 users may filter tests with the "cw310_rom" tag to
# run tests with a common bitstream available in GCP to avoid issues building
# and loading new bitstreams
```
?

Comment (dmcardle) [reply ci/scripts/check-bazel-tags.sh]: Works for me :)

If we haven't already, we should probably document this optimization at some point. Then we can just point to the doc whenever this comes up, rather than trying to re-explain why we're grouping bitstream. We'd want to update `azure-pipelines.yml`, `ci/scripts/run-fpga-cw310-tests.sh`, and this file — maybe I'm missing some.

**Comment by alphan** [comment]
@drewmacrae would you mind simplifying the commits? It looks like some are fixing up previous ones, e.g. adding cw310_other, then renaming cw310_other to cw310_rom_variant.

**Comment by drewmacrae** [comment]
> @drewmacrae would you mind simplifying the commits? It looks like some are fixing up previous ones, e.g. adding cw310_other, then renaming cw310_other to cw310_rom_variant.

It looks like they're going to be useful to keep track of this work to get the uncontroversial stuff merged.

**Pull request closed without merging**
